python/ray/autoscaler/commands.py

```diff
@@ -0,0 +1,149 @@
+    if not importer:
+        raise NotImplementedError(
+            "Unsupported provider {}".format(config["provider"]))
+
+    bootstrap_config, _ = importer()
+    config = bootstrap_config(config)
+    get_or_create_head_node(config)
+
+
+def teardown_cluster(config_file):
+    """Destroys all nodes of a Ray cluster described by a config json."""
+
+    config = yaml.load(open(config_file).read())
+
+    validate_config(config)
+    provider = get_node_provider(config["provider"], config["cluster_name"])
+    head_node_tags = {
+        TAG_RAY_NODE_TYPE: "Head",
+    }
+    for node in provider.nodes(head_node_tags):
+        print("Terminating head node {}".format(node))
+        provider.terminate_node(node)
+    nodes = provider.nodes({})
+    while nodes:
+        for node in nodes:
+            print("Terminating worker {}".format(node))
+            provider.terminate_node(node)
+        time.sleep(5)
+        nodes = provider.nodes({})
+
+
+def get_or_create_head_node(config):
+    """Create the cluster head node, which in turn creates the workers."""
+
+    provider = get_node_provider(config["provider"], config["cluster_name"])
+    head_node_tags = {
+        TAG_RAY_NODE_TYPE: "Head",
+    }
+    nodes = provider.nodes(head_node_tags)
+    if len(nodes) > 0:
+        head_node = nodes[0]
+    else:
+        head_node = None
+
+    launch_hash = hash_launch_conf(config["head_node"])
+    if head_node is None or provider.node_tags(head_node).get(
+            TAG_RAY_LAUNCH_CONFIG) != launch_hash:
+        if head_node is not None:
+            print("Terminating outdated head node {}".format(head_node))
+            provider.terminate_node(head_node)
+        print("Launching new head node...")
+        head_node_tags[TAG_RAY_LAUNCH_CONFIG] = launch_hash
+        head_node_tags[TAG_NAME] = "ray-{}-head".format(config["cluster_name"])
+        provider.create_node(config["head_node"], head_node_tags, 1)
+
+    nodes = provider.nodes(head_node_tags)
+    assert len(nodes) == 1, "Failed to create head node."
+    head_node = nodes[0]
+
+    runtime_hash = hash_runtime_conf(config["file_mounts"], config)
+
+    if provider.node_tags(head_node).get(
+            TAG_RAY_RUNTIME_CONFIG) != runtime_hash:
+        print("Updating files on head node...")
+
+        # Rewrite the auth config so that the head node can update the workers
+        remote_key_path = "~/ray_bootstrap_key.pem".format(
+            config["auth"]["ssh_user"])
+        cluster_config_path = "~/ray_bootstrap_config.yaml".format(
+            config["auth"]["ssh_user"])
+        remote_config = copy.deepcopy(config)
+        remote_config["auth"]["ssh_private_key"] = remote_key_path
+
+        # Adjust for new file locations
+        new_mounts = {}
+        for remote_path in config["file_mounts"].keys():
+            new_mounts[remote_path] = remote_path
+        remote_config["file_mounts"] = new_mounts
```

Comment:
nit: I don't think `keys()` is needed - also, below is less verbose
```
remote_config["file_mounts"] = {
    path: path for path in config["file_mounts"]}
```

Reply:
Removed keys.